### PR TITLE
Updated load functions to handle pre-computed layouts

### PR DESCRIPTION
**Pull request by ludvigla**
## Description

Current behaviour:

- Pixelator now outputs node names in the layout parquet files that are required to make sure that the order of the pre-computed layouts coordinates match the order in the graph objects imported in R. This requires modifying the LoadCellGraphs methods to handle the ordering

- Plot3DGraph expects "fr", "pmds" or "kk" as layout method. The pre-computed layouts with dim=3 have a suffix "3d", e.g. "pmds_3d" which Plot3DGraph doesn't recognise. We should update Plot3DGraph to accept any layout name and simply check that the layout is 3D.

- WriteMPX_pxl_file also need to export node names for the layout parquet files

This PR addresses the abovementioned issues. The updates are necessary to render tutorials on the software-site using pre-computed layouts.

Fixes: exe-1832

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Updated tests for ReadMPX_layouts to accept additional node names

## PR checklist:

- [x] This comment contains a description of changes (with reason).
- [x] I have performed a self-review of my own code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have checked my code and documentation and corrected any misspellings.
- [x] I have run R CMD check on the package and it passes without errors or warnings (notes can be acceptable if motivated)
- [ ] I have documented any significant changes to the code in [CHANGELOG.md](../CHANGELOG.md)

### CODE REVIEW

File: R/graph_layout_visualization.R [hunk 1]
```diff
@@ -535,7 +535,7 @@ Plot3DGraph <- function (
   cell_id,
   marker = NULL,
   assay = NULL,
-  layout_method = c("pmds", "wpmds", "fr", "kk", "drl"),
+  layout_method = "pmds",
```
Comment (vincent-van-hoef) [inline R/graph_layout_visualization.R:538]: Is there a reason to remove the vector of choices from before?

**Review by vincent-van-hoef — APPROVED** [review]

File: R/load_precomputed_layouts.R [hunk 2]
```diff
@@ -45,6 +45,13 @@ ReadMPX_layouts <- function (
 
   # Create temporary directory
   temp_layout_dir <- file.path(fs::path_temp(), "temp_layouts")
+  while (fs::dir_exists(temp_layout_dir)) {
+    err <- try(fs::dir_delete(temp_layout_dir))
+    if (inherits(err, "try-error")) {
+      warn(glue("Failed to delete temporary directory '{col_br_blue(temp_layout_dir)}'."))
+      temp_layout_dir <- file.path(fs::path_temp(), "temp_layouts", .generate_random_string())
+    }
+  }
```
Comment (maxkarlsson) [inline R/load_precomputed_layouts.R:54]: I don't fully understand this part. Do you want to remove the temporary folder if it exists in order to create a path that is not used? If so, think this could possibly be a separate utility function for readability. I would suspect we would want to use it in other places.

File: R/load_cell_graphs.R [hunk 3]
```diff
@@ -312,7 +312,16 @@ LoadCellGraphs.MPXAssay <- function (
       cg <- cg_list_full[[nm]]
```
Comment (maxkarlsson) [inline R/load_cell_graphs.R:312]: I am wondering whether `load_precomputed_layouts` should be called `load_layouts` instead? It seems like the "precomputed" is making an unnecessary assumption about how the layouts ended up in the file. Just a side point.

File: R/graph_layout_visualization.R (hunk 1 above)
Comment (ludvigla) [reply R/graph_layout_visualization.R:538]: Yes! These are the default options in `ComputeLayout`, but technically you could provide any custom method to `ComputeLayout` and give it any name you want. `Plot3DGraph` should be able to plot any layout as long as it is 3D.

Also, when running Pixelator with pre-computed layouts activated, the layout are named like this:
- "pmds" -> 2D layout
- "pmds_3d" -> 3D layout
etc.

In pixelatorR, 2D and 3D layouts have the same name. You just need decide if you want to compute with dim=2 or dim=3, but you cannot do both unless you set the `layout_name` manually. We might change this in the future, but right now we need a way to enable plotting of "pmds_3d" created with Pixelator. By removing these choices, you can simply set `layout_method = "pmds_3d"` if you have loaded the pre-computed layouts.

File: R/load_precomputed_layouts.R [hunk 4]
```diff
@@ -60,7 +67,7 @@ ReadMPX_layouts <- function (
   # Load hive-styled parquet files
   coords <- arrow::open_dataset(temp_layout_dir)
   coords <- coords %>%
-    select(any_of(c("x", "y", "z", "sample")), component, graph_projection, layout) %>%
+    select(any_of(c("name", "x", "y", "z", "sample")), component, graph_projection, layout) %>%
```
Comment (maxkarlsson) [inline R/load_precomputed_layouts.R:70]: Here we keep the name column, but in line 324 in 'load_cell_graphs' we remove it from the cellgraph@layout elements. I am not sure if relevant, but that would lead to a discrepancy between using taht function or this one directly, right?

File: R/write_pxl_file.R [hunk 5]
```diff
@@ -119,6 +119,13 @@ WriteMPX_pxl_file <- function (
 
   # Create a temporary directory
   pxl_folder <- file.path(fs::path_temp(), .generate_random_string())
+  while (fs::dir_exists(pxl_folder)) {
+    err <- try(fs::dir_delete(pxl_folder))
+    if (inherits(err, "try-error")) {
+      warn(glue("Failed to delete temporary directory '{col_br_blue(pxl_folder)}'."))
+      pxl_folder <- file.path(fs::path_temp(), .generate_random_string())
+    }
+  }
```
Comment (maxkarlsson) [inline R/write_pxl_file.R:128]: Here we are doing the same with the temporary folder, right? Yes, I think it would be nice to package it in a utility function.

File: R/write_pxl_file.R [hunk 6]
```diff
@@ -663,14 +682,14 @@ WriteMPX_pxl_file <- function (
       return(cg@counts %>% Matrix::t())
     }) %>% do.call(cbind, .)
     colnames(merged_counts) <- 1:ncol(merged_counts)
-    #merged_counts <- SeuratObject::RowMergeSparseMatrices(all_counts[[1]], all_counts[-1])
 
     merged_layouts <- lapply(names(layouts), function(layout_type) {
       layout_tbl <- layouts[[layout_type]]
+      # If the layout only has x,y columns, add a third z column of NAs
```
Comment (maxkarlsson) [inline R/write_pxl_file.R:688]: Why is this necessary?

File: R/write_pxl_file.R [hunk 7]
```diff
@@ -663,14 +682,14 @@ WriteMPX_pxl_file <- function (
...
       if (ncol(layout_tbl) == 2 & !"z" %in% names(layout_tbl)) {
         layout_tbl <- layout_tbl %>% mutate(z = NA_real_)
       }
-      layout_tbl <- layout_tbl[, c("x", "y", "z")]
+      layout_tbl <- layout_tbl %>% select(all_of(c("name", "x", "y", "z")))
```
Comment (maxkarlsson) [inline R/write_pxl_file.R:692]: Could use `any_of` instead to account for when z is missing.

**Review by maxkarlsson — COMMENTED** [review]
Looks good to me! I have some comments and suggestions.

File: R/load_precomputed_layouts.R (hunk 2 above)
Comment (ludvigla) [reply R/load_precomputed_layouts.R:54]: So what happens is that we generate a new temp folder with `file.path(fs::path_temp(), "temp_layouts", .generate_random_string())`. 

`.generate_random_string()` creates a random string with 5 letters/integers. This is just to ensure that we use an unique name to create a new empty directory. However, if we reset the seed there's a risk that we create a folder with the same name. If this folder contains layout parquet files these will also be loaded which is something we want to avoid. Usually, this shouldn't be an issue, but can happen for example when we run the tests where we reset the seed quite frequently.

Another annoying thing is that there's no guarantee that we can actually delete files in the temp folder. When we run the R CMD check on windows (github workflow) this is the case. So what I have attempted here is to create a new temporary directory name, delete it if it already exists, but if we cannot delete it we try to create a new name that is not being used. This is mainly to make sure that everything works on our github workflow (windows) but in general, the folder should be deleted if it already exists and created again. Once we have loaded the layouts, we don't need the folder anymore so it should be safe to remove.

If there's a cleaner way to do this, that would be great!

File: R/load_cell_graphs.R (hunk 3 above)
Comment (ludvigla) [reply R/load_cell_graphs.R:312]: Yeah that's a fair point. But could it be that `load_layouts` leads users to think that they are computed in R?

File: R/load_precomputed_layouts.R (hunk 4 above)
Comment (ludvigla) [reply R/load_precomputed_layouts.R:70]: That's correct. `ReadMPX_layouts` returns a list of lists of tibbles where each tibble contains an additional column with node names. The reason why its a list of list is because each component can have multiple pre-computed layouts. The order of the components in that list will probably not match the order of the components in a Seurat object generated from the same PXL file. The layout files are not ordered in any specific way in the PXL file, so they will be loaded in the order they appear.

I think this discrepancy is fine since the way layouts are stored in the Seurat object is quite different anyways.

File: R/write_pxl_file.R (hunk 5 above)
Comment (ludvigla) [reply R/write_pxl_file.R:128]: I think that's a good idea!

File: R/load_precomputed_layouts.R (hunk 4 above)
Comment (maxkarlsson) [reply R/load_precomputed_layouts.R:70]: Okay, that's fine with me then!

File: R/write_pxl_file.R (hunk 6 above)
Comment (ludvigla) [reply R/write_pxl_file.R:688]: If I remember correctly, this is how coordinates are stored in the PXL files. If you compute pmds 2D, you still get a z column with empty values. But I need to double check this with Johan. Unfortunately, I don't have access to any PXL files with 2D layouts.

File: R/write_pxl_file.R (hunk 7 above)
Comment (ludvigla) [reply R/write_pxl_file.R:692]: See my answer above. I don't think that will match how the layouts should be stored in the PXL files.

File: R/load_cell_graphs.R (hunk 3 above)
Comment (ludvigla) [reply R/load_cell_graphs.R:312]: I will update the argument name to `load_layouts`!

Comment (ludvigla) [reply R/load_cell_graphs.R:311]: Now the script name has changed from `load_precomputed_layouts.R` to `load_layouts.R` and the argument passed to `LoadCellGraphs` is now called `load_layouts`.

File: R/write_pxl_file.R (hunk 5 above)
Comment (ludvigla) [reply R/write_pxl_file.R]: Added an internal function `.create_unique_temp_dir` to `utils.R` that replaces these code blocks.
